lib/AST/ASTMangler.cpp

```diff
@@ -2167,9 +2167,29 @@ ASTMangler::getSpecialManglingContext(const ValueDecl *decl,
   if (isa<TypeDecl>(decl)) {
     if (auto *clangDecl = cast_or_null<clang::NamedDecl>(decl->getClangDecl())){
       bool hasNameForLinkage;
-      if (auto *tagDecl = dyn_cast<clang::TagDecl>(clangDecl))
+      if (auto *tagDecl = dyn_cast<clang::TagDecl>(clangDecl)) {
         hasNameForLinkage = tagDecl->hasNameForLinkage();
-      else
+        // Clang does not always populate the fields that determine if a tag
+        // decl has a linkage name in the ways we expect. To support differing
+        // definitions of CF_OPTIONS, we check if the tag decls is anonymous and
+        // backed by an unavailable typedef.
+        if (!hasNameForLinkage) {
```

Comment:
This is a little weird. I want to make the Clang AST look like the Objective-C definition so that the mangled name will be the same, but I don't want the mutation to persist beyond after the mangler is done with it. More specifically, I want the anonymous enum to no longer be anonymous and instead have the backing typedef's name.

I considered making a copy of the node, but that seems pretty wasteful. I'd have to copy at least once per NS_OPTION use and keep them around somewhere. Instead I've made this RAII container that just clears the anonymous enum's name after the client is done with it.